### DISCUSSION TITLE
Validating Enum by name

### DISCUSSION THREAD

**Discussion by al-dpopowich** [pydantic/pydantic#2980] [Feature Request]
I have not been able to find this in a Discussion or Issue...

If I have a python Enum:

```python
import enum

class StateAbbrev(enum.Enum):
   """State Abbreviations"""
   NY = "New York"
   MA = "Massachusetts"
   # etc
```

I can use this with SQLAlchemy...

```python
from sqlalchemy import Table, Column, Enum # etc
User = Table(
   'myuser',
   ...
   Column('state', Enum(StateAbbrev))
   ...
)
```

On saving, sqlalchemy uses *name* **NOT** the *value*.  That is, sqlalchemy will store "NY", not "New York".  Similarly, via my REST API, I want to accept the "codes" not the (possibly) long strings.  This is a disconnect from pydantic which validates by value.

I have seen answers to similar questions on stackoverflow suggest using the same value as name, e.g:

```python
class Foo(enum.Enum):
   A = "A"
   B = "B"
```

But this is not practical when you want to offer your Enum as a CODE=>HUMAN_READABLE mapping.

In short, what I want is a Config option, e.g., `validate_enum_by_name` (for backward compatibility, default to `False`):

```python
# in validators.py
def enum_member_validator(v: Any, field: 'ModelField', config: 'BaseConfig') -> Enum:
    enum = field.type_
    try:
        enum_v = enum[v] if config.validate_enum_by_name else enum(v)
    except (KeyError, ValueError):
        # enum should be an enum, so will be iterable
        raise errors.EnumMemberError(enum_values=list(enum))
    return enum_v.value if config.use_enum_values else enum_v
```

**Comment by al-dpopowich** [comment]
My workaround is to create a custom type that is a subclass of `enum.Enum`.  With `__get_validators__()` defined it overrides the default validator for an Enum.  Still, it would be nice to have the config option I suggest above.  Or perhaps note the following in documentation:

```python
from enum import Enum

import pydantic.json

class EnumByName(Enum):
   """A custom Enum type for pydantic to validate by name.
   """
   # The reason for this class is there is a disconnect between how
   # SQLAlchemy stores Enum references (by name, not value; which is
   # also how we want our REST API to exchange enum references, by
   # name) and Pydantic which validates Enums by value.  We create a
   # Pydantic custom type which will validate an Enum reference by
   # name.

   # Ugliness: we need to monkeypatch pydantic's jsonification of Enums
   pydantic.json.ENCODERS_BY_TYPE[Enum] = lambda e: e.name

   @classmethod
   def __get_validators__(cls):
      # yield our validator
      yield cls._validate

   @classmethod
   def __modify_schema__(cls, schema):
      """Override pydantic using Enum.name for schema enum values"""
      schema['enum'] = list(cls.__members__.keys())

   @classmethod
   def _validate(cls, v):
      """Validate enum reference, `v`.  

      We check:
        1. If it is a member of this Enum
        2. If we can find it by name.
      """
      # is the value an enum member?
      try:
         if v in cls:
            return v
      except TypeError:
         pass

      # not a member...look up by name
      try:
         return cls[v]
      except KeyError:
         name = cls.__name__
         expected = list(cls.__members__.keys())
         raise ValueError(f'{v} not found for enum {name}. Expected one of: {expected}')
```

**Reply by neilrla** [reply]
Thanks for the workaround, I've been looking for the right way to do this for some time.

**Reply by anilsg** [reply]
Thanks for the workaround, I haven't tried it yet but I've been looking for how to do this.

**Reply by neilrla** [reply]
I can see now that the workaround is intended to change the behaviour for *all* enums. I don't think this is going to be the norm. Certainly for my use case, it's not a matter of me just having a different style or preference, I have a particular requirement that I need to fulfill for a particular Enum in my `BaseModel`, but not for the others. I'm trying to update this workaround to just apply to one particular field. I imagine this requirement could arise for a number of Enums in a `BaseModel` as well, but not necessarily as a rule for all.

**Reply by neilrla** [reply]
Actually I need to create a `float` Enum dynamically, validating on Enum names, not values. I can do that using `Enum()` (as a callable) but now I've got to find how to bind class methods to the Enum I get from there for `__get_validators__` and I've no idea how to introduce the schema information:
https://stackoverflow.com/questions/75587442/validate-pydantic-dynamic-float-enum-by-name-with-openapi-description

**Comment by anilsg** [comment]
Just to clarify the need for this, this is actually quite a common use case.

The use case is when I want to accept coded values for a field (classic Enum use-case) but internally I want the exact machine equivalent that's stored in the Enum value. And again, when returning a response, the user doesn't want to / doesn't need to / shouldn't receive the internal values which may be an implementation detail anyway. They want to get the same codes back.

From [one related question I just found on Stack Overflow](https://stackoverflow.com/a/75586485):

```
class Group(enum.Enum):
    user = 0
    manager = 1
    admin = 2
```

The user should provide "user", "manager", etc in the request. The user should see "user", "manager", etc in the serialised JSON response. Internally I want to see the `int` when accessing an attribute of a `BaseModel` that has this `Group` enum as it's type. If I have to use `model.field.value` to get the `int`, that's fine, but I should be able to use an Enum to describe the parameter to the OpenAPI spec, and get the Enum names to be used instead of the values.

I would image it would be right to target being able to choose whether names or values are used on both input and output, although `Config` `json_encoders` would presumably deal with all output scenarios.

**Comment by tscheburaschka** [comment]
Hi,
I'm sorry to hook into this discussion with a perhaps not exactly similar question. If this is the wrong place to ask, then I can move this to a new discussion.
So with pydantic 2 (almost) fresh on our table, I'm trying to achieve enum deserialisation (i.e. validation) by name with the more flexible type annotations. It is pretty straightforward to implement a BeforeValidator for a specific enum, like so:

```python
from enum import IntEnum
from typing import Annotated

import pytest
from pydantic import BaseModel, BeforeValidator, ValidationError


class LogLevel(IntEnum):
    DEBUG = 1
    INFO = 2


def log_level_by_name(v: str) -> LogLevel:
    try:
        return LogLevel[v]
    except KeyError:
        raise ValueError("invalid value")


LogLevelByName = Annotated[LogLevel, BeforeValidator(log_level_by_name)]


class AppConfig(BaseModel):
    log_level: LogLevelByName


def test_enum():
    c = AppConfig.model_validate({"log_level": "DEBUG"})
    assert c.log_level is LogLevel.DEBUG

    with pytest.raises(ValidationError):
        c = AppConfig.model_validate({"log_level": 1})

```
Obviously this is not very usable as it requires a new validator function for each enum type.
For a generic solution, the `BeforeValidator` function would need access to the annotated type. I speculate that there should be a way to get to it, I just can't find it.
Any help is appreciated. Thank you.

**Reply by anilsg** [reply]
@tscheburaschka I just wanted to check did you refer to the solution provided here: https://stackoverflow.com/questions/75587442/validate-pydantic-dynamic-float-enum-by-name-with-openapi-description ?
I haven't looked at your requirements and compared but that solution is a generic solution and you may be able to modify it for your purpose. One variant uses a decorator and I'm thinking maybe you could use a decorator to pass in the the enum type that you need.

**Reply by tscheburaschka** [reply]
Thank you for the hint, @anilsg. I had seen this SO post before but honestly found it a too big weapon to use and was hoping that the new type annotation possibilities of pydantic 2 allow a more elegant solution.
On a second look I can see, that the decorator solution satisfies all my requirements, and in absence of more enlightenment about type annotations (wherever this may come from) I will go with this option.
Thank you for re-emphasizing the decorator.

**Reply by tscheburaschka** [reply]
FYI: The [class decorator solution](https://stackoverflow.com/a/75597031) mentioned above breaks in multiple ways with pydantic 2. 
E.g.:
```
    def __enum_name_validator__(v: Any, field: ModelField) -> Any:
>       assert isinstance(field.type, EnumMeta)
E       AttributeError: 'pydantic_core._pydantic_core.ValidationInfo' object has no attribute 'type'
```

**Reply by anilsg** [reply]
Ah yes, indubitably. Thanks for raising that @tscheburaschka . If I need that again (not unlikely), then when I have to fix it up I will update the SO question, probably with an additional answer, since the existing answers are big enough already, as you pointed out.

Meanwhile, if you make some changes, or even create a related-but-not-identical resource that's in scope, I would really appreciate your answer there too. I feel SO is sinking into the mire these days, not just because of volumes of poor content, but also because of egotistic moderation and interference, but it still remains a good source of answers and some questions (like this one) are actually documentation locations for some of us poor souls.

I still find it interesting that under both Python `enum` and Pydantic use of them in `BaseModel`, no-one seems to have caught the use-case of name evaluated enums. I have a client who is only allowed to use symbolic names for the field, but internally I have implementation related concrete values that are subject to change. Happens all the time. It should really be the default approach.

We go to all the length of constructing an `enum` with *pairs* of values (we keep _two_ values instead of one) and then we fail to de-couple them properly. It's almost like we miss the point of the `enum` implementation in the first place. We have a list of names with associated values and then insist on users always referencing the values. It's like creating a `dict` and then expecting callers to always refer to `values` instead of `keys`.

Then we have to build these "weapons" to make it work how you would expect it to be designed in the first place.

**Reply by tscheburaschka** [reply]
I have not enough reputation to comment on SO, so I'll post here or at least add a reference in any case.

I agree, that the `enum`-implementation in Python has some flaws in its core design. As a solution I separated the mapping from the enumeration and are now using only the bare string-to-string type enums and map those to business objects in a second step.

**Reply by anilsg** [reply]
I was thinking you might post an actual answer there @tscheburaschka, not a comment (not enough room anyway), but I'll take whatever I can get.
I have this alternative `enum` customisation too.
It's probably not capable enough for your requirements (I guess) not being generic but it may give you ideas.
It follows `str`/`str` `enum` pattern but passes two arguments in during creation, storing one of them as an additional attribute that you can access internally:

```
import enum
from decimal import Decimal

class Frequency(str, enum.Enum):

    duration: Decimal

    def __new__(cls, value: str, duration: Decimal = Decimal("0")) -> "Frequency":
        obj = str.__new__(cls, value)
        obj._value_ = value
        obj.duration = duration
        return obj

    ANNUALLY = ("annually", Decimal("365"))
    QUARTERLY = ("quarterly", Decimal("365") / 4)
    MONTHLY = ("monthly", Decimal("365") / 12)
    MONTH30 = ("month30", Decimal("30"))
    FORTNIGHTLY = ("fortnightly", Decimal("365") / 26)
    WEEKLY = ("weekly", Decimal("365") / 52)
```

**Comment by dkbarn** [comment]
When you consider the fact that Python offers the `enum.auto()` function for dynamically allocating values to enum members, it is clear that the name can be more important than the value.

```python
class MyEnum(enum.Enum):
    foo = enum.auto()
    bar = enum.auto()
    baz = enum.auto()
```
I would obviously want this enum to be serialized and deserialized by name rather than by value. Use of `enum.auto()` implies the value is arbitrary -- you only care that each member has a unique one, but not necessarily what it is. If I were to later insert a new member in between "foo" and "bar", the values are going to change.

It would be great if Pydantic recognized that enums with a static name and dynamic value are actually very common, and would provide built-in support for this scenario so we didn't have to resort to implementing our own custom validators and serializers.

**Reply by WilliamStam** [reply]
add a pwus won on this. 

```
class MyEnum(Enum)
   FOO = "FOO"
   BAR = "BAR"

   def get_value(self): #psudo code
      match self.value:
         case MyEnum.FOO:
             return "foo is the first" 
```
is a pain :(

(the having to define name and value the same and then add separate logic to get the "real" value)

**Reply by baarney** [reply]
Or without even bothering with  `enum.auto()`:
```
Color = Enum('Color', ['RED', 'GREEN', 'BLUE'])
```
Most of my enums are like this - I'd love a global config setting to just use name for all serialisation, validation, OpenAPI docs etc.

**Comment by WilliamStam** [comment]
im playing around with something like this 

```
orderType = NewType('Order', enum.Enum("order", [(g.name, g.name) for g in ModelNameEnum]))
```
but then to transfer it back to the origional enum `ModelNameEnum[val]` but its been hurting my brain for a few days now

**Comment by baarney** [comment]
I'm converting an app to FastAPI and for pretty much all my usage I consider the value of an enum to be an internal thing that should never appear in the API calls. Two examples to illustrate:

1. Simple enum: I don't care about the value. In fact, for these cases I don't even care that there is a separate value - that is just something Python adds under the hood for its own use:
```
Color = Enum('Color', ['RED', 'GREEN', 'BLUE'])
```
If I later switch that to:
```
Color = Enum('Color', ['CYAN', 'MAGENTA', 'YELLOW', 'BLACK'])
```
I want the API clients to have to change their calls - if they just pass 1, 2, 3, then they may get the wrong colors with no indication of errors.

2. Complex enum: say I have some boxes for my orders:
```
class Box(Enum):
    SMALL = {'length': 100, 'width': 20, 'depth': 10}
    MEDIUM = {'length': 200, 'width': 40, 'depth': 20}
    LARGE = {'length': 400, 'width': 60, 'depth': 60}
    
app = FastAPI()

@app.get("/order")
async def get_box(box: Box):
    return box
```
All I want in the API is:
```
.../order?box=SMALL
```
With the value implementation it appears I have to pass some mangled version of the entire value dictionary.

Of course I can fix all these problems for each individual case, but a global config flag to just use name instead of value would be so much simpler.

I had a quick hack at `pydantic/_internal/_std_types_schema.py` and this seems to fix the model validation / generation:
```
69a70,74
>     # add to / get from ConfigDict
>     use_enum_name = True
> 
>     def nv(elem):
>         return elem.name if use_enum_name else elem.value
88c93
<         expected = repr(cases[0].value)
---
>         expected = repr(nv(cases[0]))
90c95
<         expected = ', '.join([repr(case.value) for case in cases[:-1]]) + f' or {cases[-1].value!r}'
---
>         expected = ', '.join([repr(nv(case)) for case in cases[:-1]]) + f' or {nv(cases[-1])!r}'
94c99,102
<             enum_field = enum_type(__input_value)
---
>             if use_enum_name:
>                 enum_field = __input_value if isinstance(__input_value, enum_type) else enum_type[__input_value]
>             else:
>                 enum_field = enum_type(__input_value)
109c117,120
<     if issubclass(enum_type, int):
---
>     if use_enum_name:
>         lax = to_enum_validator
>         strict = core_schema.json_or_python_schema(json_schema=to_enum_validator, python_schema=strict_python_schema)
>     elif issubclass(enum_type, int):
```
Certainly not tested all cases, and this doesn't fix serialisation, but seems to suggest it would not be a huge change.

**Comment by valery-iwanofu** [comment]
My solution for this case:
```python
import enum
from typing import Any
from pydantic import GetCoreSchemaHandler
from pydantic_core import core_schema


def pydantic_enum[E: enum.Enum](enum_cls: type[E]) -> type[E]:
    def __get_pydantic_core_schema__(cls: type[E], source_type: Any, handler: GetCoreSchemaHandler):
        assert source_type is cls
        
        def get_enum(value: Any, validate_next: core_schema.ValidatorFunctionWrapHandler):
            if isinstance(value, cls):
                return value
            else:
                name: str = validate_next(value)
                return enum_cls[name]

        def serialize(enum: E):
            return enum.name
        
        expected = [member.name for member in cls]
        name_schema = core_schema.literal_schema(expected)
        
        return core_schema.no_info_wrap_validator_function(
            get_enum, name_schema, 
            ref=cls.__name__,
            serialization=core_schema.plain_serializer_function_ser_schema(serialize)
        )
    
    setattr(enum_cls, '__get_pydantic_core_schema__', classmethod(__get_pydantic_core_schema__))
    return enum_cls

```

Usage example:
```python
from pydantic import BaseModel, ValidationError

@pydantic_enum
class MyEnum(enum.Enum):
    biba = enum.auto()
    boba = enum.auto()

class Owner(BaseModel):
    value: MyEnum

assert Owner.model_validate({
    'value': 'biba'
}).value == MyEnum.biba
assert Owner.model_validate({
    'value': MyEnum.boba
}).value == MyEnum.boba

try:
    Owner.model_validate({
        'value': 1
    })
    assert False
except ValidationError:
    pass

assert Owner(value=MyEnum.biba).model_dump(mode='json') == {'value': 'biba'}
```

Unfortunately, it won't work with FastAPI response because of [this](https://github.com/tiangolo/fastapi/blob/2afbdb3a44b6b5af0d6b07d435f4f7c7d9bea1bc/fastapi/encoders.py#L256)

**Reply by eltoder** [reply]
@valery-iwanofu it does work with FastAPI if you annotate your return type.

**Reply by gsiddhant159** [reply]
I also had to add context info to the serialization function to allow dumping the enum instance when used for database operations

```python
class SerializationContext(enum.StrEnum):
    FASTAPI = enum.auto()
    SQLA = enum.auto()
```

```python
def serialize(enum: E, info: SerializationInfo):
    if info.context and info.context.get("mode") == SerializationContext.SQLA:
        return enum

    return enum.name.lower()
```

Is there a better way to handle this?

**Comment by baarney** [comment]
This also means that Pydantic does not support enum aliases.
```py
class Shape(Enum):
    SQUARE = 2
    DIAMOND = 1
    CIRCLE = 3
    ALIAS_FOR_SQUARE = 2
```
If an API returns `ALIAS_FOR_SQUARE` the caller gets `2` - if they send this back in another call it gets converted to `SQUARE`. There is no way for a client to pass `ALIAS_FOR_SQUARE` in an API call.

The OpenAPI Schema also contains the duplicate value:
```
Shape: {
  type: "integer",
  enum: [2, 1, 3, 2],
  title: "Shape"
},
```

**Comment by gerhardadler** [comment]
My solution to the problem is this:
```python
class EnumByName:
    def __init__(self, *, ignore_case: bool = True):
        self.ignore_case = ignore_case

    def __get_pydantic_core_schema__(self, enum_cls: type[Enum], _handler: GetCoreSchemaHandler):
        name_enum = Enum('name_enum', {member.name: member.name for member in enum_cls})
        name_enum = cast(type[Enum], name_enum)

        def enum_or_name(value: Enum | str) -> Enum:
            if isinstance(value, str):
                if not self.ignore_case:
                    try:
                        return enum_cls[value]
                    except KeyError:
                        raise ValueError(f'Enum name not found: {value}')
                try:
                    return next(
                        member for member in enum_cls if member.name.lower() == value.lower()
                    )
                except StopIteration:
                    raise ValueError(f'Enum name not found: {value}')
            elif isinstance(value, enum_cls):
                return value
            raise ValueError(f'Expected enum member or name, got {type(value).__name__}: {value}')

        return core_schema.no_info_plain_validator_function(
            enum_or_name,
            json_schema_input_schema=core_schema.enum_schema(
                enum_cls, list(name_enum.__members__.values())
            ),
            ref=enum_cls.__name__,
            serialization=core_schema.plain_serializer_function_ser_schema(lambda e: e.name),
        )


class MyEnum(Enum):
    A = 1
    B = 2
    C = 3


class MyModel(BaseModel):
    my_enum: Annotated[MyEnum, EnumByName()]
```

It has a few advantages:
* It does not modify the original Enum.
* The JSON schema is identical to a normal Pydantic Enum, apart from using the names instead.
* Works with FastAPI

**Reply by baarney** [reply]
Love it! Needed one small addition to work for me...

For some reason this wasn't picking up the schema in the output type - I was getting `Cannot generate a JsonSchema for core_schema.PlainValidatorFunctionSchema` when generating the Open API schema. Adding a `return_schema` to the `serialization` seemed to fix this.
```
serialization=core_schema.plain_serializer_function_ser_schema(
    lambda e: e.name,
    return_schema=core_schema.literal_schema(list(enum_cls.__members__.keys()))
)
```

If you also use enums in path/query parameters then this appears to generate two separate OpenAPI schemas, one for input and one for output, instead of sharing a single one as the built-in enums do. This could possibly have a knock-on effect if something is using the schema to generate code, but is not a problem for me.
```
"MyEnum-Input": {
  "type": "string",
  "enum": [
    "A",
    "B",
    "C"
  ],
  "title": "MyEnum"
},
"MyEnum-Output": {
  "type": "string",
  "enum": [
    "A",
    "B",
    "C"
  ]
},
```

**Comment by silviogutierrez** [comment]
I generally want this behavior globally, not opt-in. This seems to work well:

```python3
class ByNameEnumGenerateSchema(GenerateSchema):
    def match_type(self, obj: Any) -> core_schema.CoreSchema:
        if inspect.isclass(obj) and issubclass(obj, enum.Enum):

            def get_enum(
                value: Any, validate_next: core_schema.ValidatorFunctionWrapHandler
            ) -> Any:
                if isinstance(value, obj):
                    return value
                else:
                    name: str = validate_next(value)
                    return obj[name]

            def serialize(
                value: enum.Enum, info: core_schema.SerializationInfo
            ) -> str | enum.Enum:
                if info.mode == "json":
                    return value.name
                return value

            expected = [member.name for member in obj]
            name_schema = core_schema.literal_schema(expected)

            return core_schema.no_info_wrap_validator_function(
                get_enum,
                name_schema,
                ref=obj.__qualname__,
                serialization=core_schema.plain_serializer_function_ser_schema(
                    serialize, info_arg=True
                ),
            )

        return super().match_type(obj)
```


```python3
class CustomModel(BaseModel):
    model_config = ConfigDict(
        schema_generator=ByNameEnumGenerateSchema, defer_build=True
    )
```


Then just use `CustomModel` everywhere in my app. Tested against `pydantic==2.7.4`

**Comment by rhpreston** [comment]
If you only need to accept enum names on validation keeping serialization the same, you can do this:

```python
from enum import Enum
from functools import partial
from typing import Annotated, Any, Type

from pydantic import BaseModel, BeforeValidator


def accept_enum_name(enum: Type[Enum], value: Any) -> Any:
    if isinstance(value, str) and value in enum.__members__:
        return enum[value]
    else:
        return value

class EnumField(Enum):
    ENUM_NAME = "enum value"

class Model(BaseModel):
    enum_field: Annotated[EnumField, BeforeValidator(partial(accept_enum_name, EnumField))]
```

Still trying to understand how to apply the behavior globally as the previous comment shows.

Edit: I've found the below pattern is quite nice.

```python
from enum import Enum
from typing import Annotated, Any, Type

from pydantic import BeforeValidator

def accept_enum_name(enum: Type[Enum]) -> BeforeValidator:
    def validator(value: Any) -> Any:
        if isinstance(value, str) and value in enum.__members__:
            return enum[value]
        else:
            return value
    return BeforeValidator(validator)

class EnumField(Enum):
    ENUM_NAME = "enum value"

class Model(BaseModel):
    enum_field: Annotated[EnumField, accept_enum_name(EnumField)]
```

**Comment by baarney** [comment]
Is there enough here now to move this from a discussion back to a feature request?

- Add a boolean configuration option: "serialise_enum_by_name" or similar, default False.
- If it is False: no change to serialisation / schemas
- If it is True: serialisation / de-serialisation uses the enum name, and schemas contain a list of enum names

**Reply by silviogutierrez** [reply]
I would hope so. Now GenerateSchema overriding is completely gone, so my previous solution doesn't work. See here: https://github.com/pydantic/pydantic/discussions/2980#discussioncomment-12985212

For those interested in forcing this, here is a **grotesque** monkey patch:

```
from __future__ import annotations

import enum
from typing import Any

from pydantic._internal._generate_schema import GenerateSchema
from pydantic_core import core_schema


# Pydantic 2.10+ deprecated schema_generator and 2.12 ignores it entirely.
# Monkey-patch GenerateSchema._enum_schema so enums validate/serialize by
# member NAME (e.g. "FOO") instead of by value (e.g. "Foo").
def _enum_schema_by_name(
    self: GenerateSchema, enum_type: type[enum.Enum]
) -> core_schema.CoreSchema:
    def get_enum(
        value: Any, validate_next: core_schema.ValidatorFunctionWrapHandler
    ) -> Any:
        if isinstance(value, enum_type):
            return value
        name: str = validate_next(value)
        return enum_type[name]

    def serialize(
        value: enum.Enum, info: core_schema.SerializationInfo
    ) -> str | enum.Enum:
        if info.mode == "json":
            return value.name
        return value

    expected = list(enum_type.__members__.keys())
    name_schema = core_schema.literal_schema(expected)

    return core_schema.no_info_wrap_validator_function(
        get_enum,
        name_schema,
        ref=enum_type.__qualname__,
        serialization=core_schema.plain_serializer_function_ser_schema(
            serialize, info_arg=True
        ),
    )


GenerateSchema._enum_schema = _enum_schema_by_name  # type: ignore[method-assign]
```

**Comment by baarney** [comment]
Created a feature request: https://github.com/pydantic/pydantic/issues/13192
